### PR TITLE
PartsSplitter invalid ranges for the same part

### PR DESCRIPTION
**Pull request by kitaisreal**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fixed a bug in parallel optimization for queries with `FINAL`, which could give an incorrect result in rare cases. 

PartsSplitter can produce invalid read ranges when primary key value for end range is equal to primary key value for another range start in the same part.

### CODE REVIEW

**Comment by robot-clickhouse-ci-2** [comment]
*This is an automated comment for commit 8a539f222eb0e9bca039d538ff293d867484fbfd with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/60041/8a539f222eb0e9bca039d538ff293d867484fbfd/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/60041/8a539f222eb0e9bca039d538ff293d867484fbfd/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Comment by kitaisreal** [comment]
@nickitat could you please check updated implementation of functionality that splits ranges into layers. Before it was also broken, because our sort comparator did not provide strict weak ordering.

File: src/Processors/QueryPlan/PartsSplitter.cpp [hunk 1]
```diff
@@ -174,23 +176,20 @@ struct PartsRangesIterator
         else if (compare_result == 1)
             return false;
 
-        if (part_index == other.part_index)
+        if (event == other.event)
```
Comment (nickitat) [inline src/Processors/QueryPlan/PartsSplitter.cpp:179]: could you pls give an example when strict weak ordering wasn't held before?
maybe there is some less radical way to fix other than allow overlapping ranges?

**Comment by nickitat** [comment]
of course it's hard to say if everything is finally 100% correct. I think the best "review" would be to add as much checks as possible (even relatively heavy that would be done only in debug build).
and in principle we could run this logic even for ordinary merge tree and it shouldn't break any test (except when we specifically check pipeline or smth alike). under a specific setting made only for tests. wdyt?

**Comment by kitaisreal** [comment]
In LLVM there is algorithm with complexity O(n^2) https://github.com/llvm/llvm-project/blob/main/libcxx/include/__debug_utils/strict_weak_ordering_check.h#L28 that check strict weak ordering for range and comparator, but it did not provide debug message for which elements comparator is broken.

With such function, we can check strict weak ordering for range and also get nice debug message for which elements comparator is broken with details. Implementation was taken from https://github.com/yugr/sortcheckxx/blob/master/include/sortcheck.h#L127. Note that algorithm is O(n^3) so it is not possible to always check range using this algorithm, even in Debug mode.

```
template <typename RandomAccessIterator, typename Compare>
inline void checkRangeForStrictWeakOrdering(RandomAccessIterator first, RandomAccessIterator last, Compare comp)
{
    static constexpr int8_t SORTCHECK_LESS = -1;
    static constexpr int8_t SORTCHECK_EQUAL = 0;
    static constexpr int8_t SORTCHECK_GREATER = 1;

    const size_t n = size_t(last - first);
    std::vector<std::vector<int8_t>> cmp(n, std::vector<int8_t>(n, 0));

    auto sortcheck_to_string = [](int8_t value)
    {
        if (value == SORTCHECK_EQUAL)
            return "equivalent";

        return value == SORTCHECK_LESS ? "less" : "greater";
    };

    for (size_t i = 0; i < n; ++i)
        for (size_t j = 0; j < n; ++j)
            cmp[i][j] = comp(*(first + i), *(first + j)) ? SORTCHECK_LESS : SORTCHECK_GREATER;

    for (size_t i = 0; i < n; ++i)
    {
        for (size_t j = 0; j <= i; ++j)
            if (cmp[i][j] == SORTCHECK_GREATER && cmp[j][i] == SORTCHECK_GREATER)
                cmp[i][j] = cmp[j][i] = SORTCHECK_EQUAL;
    }

    for (size_t i = 0; i < n; ++i)
        if (cmp[i][i] != SORTCHECK_EQUAL)
            throw Exception(ErrorCodes::LOGICAL_ERROR, "Irreflexive comparator at position {}", i);

    for (size_t i = 0; i < n; ++i)
    {
        for (size_t j = 0; j < i; ++j)
            if (cmp[i][j] != -cmp[j][i])
                throw Exception(ErrorCodes::LOGICAL_ERROR, "Non-asymmetric comparator at positions {} and {}", i, j);
    }

    for (size_t i = 0; i < n; ++i)
    {
        for (size_t j = 0; j < i; ++j)
        {
            for (size_t k = 0; k < n; ++k)
            {
                if (cmp[i][j] == cmp[j][k] && cmp[i][k] != cmp[i][j])
                {
                    throw Exception(ErrorCodes::LOGICAL_ERROR,
                        "Non-transitive {0} comparator at positions {1}, {2} and {3}. {1} is {0} than {2} and {2} is {0} than {3}, but {1} is {4} than {3}",
                        sortcheck_to_string(cmp[i][j]), i, j, k, sortcheck_to_string(cmp[i][k]));
                }
            }
        }
    }
}
```
If we have such comparator:
```
[[maybe_unused]] bool operator<(const PartsRangesIterator & other) const
{
    int compare_result = compareValues(value, other.value);
    if (compare_result == -1)
        return true;
    else if (compare_result == 1)
        return false;

    if (part_index == other.part_index)
    {
        /// Within the same part we should process events in order of mark numbers,
        /// because they already ordered by value and range ends have greater mark numbers than the beginnings.
        /// Otherwise we could get invalid ranges with the right bound that is less than the left bound.
        const auto ev_mark = event == EventType::RangeStart ? range.begin : range.end;
        const auto other_ev_mark = other.event == EventType::RangeStart ? other.range.begin : other.range.end;
        return ev_mark < other_ev_mark;
    }

    if (event == other.event)
        return part_index < other.part_index;

    // Start event always before end event
    return event < other.event;
}
```

We can run test `02990_parts_splitter_invalid_ranges.sql` with this diagnostic enabled, I modified output a bit to add indexes.
```
2024.02.21 17:46:45.012585 [ 899192 ] {28b60cc7-6209-42e7-800d-cda2c6d8f5ea} <Debug> PartsSplitter: Parts ranges after sort

0 Part index 1 event Range Start range begin 0 end 1 value (1, 1704468357009)
1 Part index 2 event Range Start range begin 0 end 1 value (1, 1704470704762)
2 Part index 0 event Range Start range begin 0 end 1 value (1, 1704472004759)
3 Part index 0 event Range End range begin 0 end 1 value (3, 1704153600000)
4 Part index 0 event Range Start range begin 2 end 3 value (3, 1704153600000)
5 Part index 1 event Range Start range begin 2 end 3 value (3, 1704153600000)
6 Part index 2 event Range End range begin 0 end 1 value (3, 1704153600000)
7 Part index 2 event Range Start range begin 2 end 3 value (3, 1704153600000)
8 Part index 1 event Range End range begin 0 end 1 value (3, 1704153600000)
9 Part index 0 event Range End range begin 2 end 3 value (5, 1700161822134)
10 Part index 1 event Range End range begin 2 end 3 value (5, 1701458520878)
11 Part index 2 event Range End range begin 2 end 3 value (5, 1702609856302)

2024.02.21 17:46:45.012842 [ 899192 ] {28b60cc7-6209-42e7-800d-cda2c6d8f5ea} <Fatal> : Logical error: 'Non-transitive greater comparator at positions 4, 3 and 5. 4 is greater than 3 and 3 is greater than 5, but 4 is less than 5'.
```

File: src/Processors/QueryPlan/PartsSplitter.cpp (hunk 1 above)
Comment (kitaisreal) [reply src/Processors/QueryPlan/PartsSplitter.cpp:179]: Added comment in pull request. We need to have `Start event` always before `End event`. We cannot change this property only for ranges inside single part.

**Comment by kitaisreal** [comment]
> of course it's hard to say if everything is finally 100% correct. I think the best "review" would be to add as much checks as possible (even relatively heavy that would be done only in debug build). and in principle we could run this logic even for ordinary merge tree and it shouldn't break any test (except when we specifically check pipeline or smth alike). under a specific setting made only for tests. wdyt?

I added setting `merge_tree_read_split_ranges_into_intersecting_and_non_intersecting_fault_probability` and necessary debug checks.
I did not add a lot of debug checks into `splitIntersectingPartsRangesIntoLayers`, what things do you think we need to check ? I just do not want to add a lot of code that will be hard to backport.

**Review by nickitat — APPROVED** [review]

**Comment by kitaisreal** [comment]
Hello @nickitat I decided to split actual fix and fault injection into two separate pull requests, so we can backport fix without too many changes.
Here is fault injection pull request https://github.com/ClickHouse/ClickHouse/pull/60548. It seems that we are safe to merge actual fix. Right now I enabled `merge_tree_read_split_ranges_into_intersecting_and_non_intersecting_fault_probability` to 1.0 for test purposes, and there are no failed tests, except ones that rely on pipeline in explain or `read_rows` amount. 
Could you please help with https://s3.amazonaws.com/clickhouse-test-reports/60041/dbc7631ab19739998d90f3f7bb17ff964d6ba118/bugfix_validation.html check in this pull request ? It seems like it is broken:
```
2024.03.01 02:40:12.240146 [ 664 ] {} <Warning> Context: Delay accounting is not enabled, OSIOWaitMicroseconds will not be gathered. Check /proc/sys/kernel/task_delayacct
2024.03.01 02:40:12.247047 [ 664 ] {} <Error> Application: Code: 115. DB::Exception: Unknown setting max_postpone_time_for_failed_mutations_ms: in MergeTree config. (UNKNOWN_SETTING), Stack trace (when copying this message, always include the lines below):

0. ./build_docker/./src/Common/Exception.cpp:96: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000c7faf9b in /usr/lib/debug/usr/bin/clickhouse.debug
1. DB::Exception::Exception<String>(int, FormatStringHelperImpl<std::type_identity<String>::type>, String&&) @ 0x0000000007213a83 in /usr/lib/debug/usr/bin/clickhouse.debug
2. ./build_docker/./src/Core/BaseSettings.cpp:0: DB::BaseSettingsHelpers::throwSettingNotFound(std::basic_string_view<char, std::char_traits<char>>) @ 0x0000000010516fcb in /usr/lib/debug/usr/bin/clickhouse.debug
3. ./build_docker/./src/Storages/MergeTree/MergeTreeSettings.cpp:24: DB::BaseSettings<DB::MergeTreeSettingsTraits>::set(std::basic_string_view<char, std::char_traits<char>>, DB::Field const&) @ 0x0000000012532153 in /usr/lib/debug/usr/bin/clickhouse.debug
4. ./build_docker/./src/Storages/MergeTree/MergeTreeSettings.cpp:37: DB::MergeTreeSettings::loadFromConfig(String const&, Poco::Util::AbstractConfiguration const&) @ 0x0000000012544791 in /usr/lib/debug/usr/bin/clickhouse.debug
5. ./build_docker/./src/Interpreters/Context.cpp:0: DB::Context::getMergeTreeSettings() const @ 0x0000000010c77e8a in /usr/lib/debug/usr/bin/clickhouse.debug
6. ./src/Core/SettingsFields.h:48: DB::Server::main(std::vector<String, std::allocator<String>> const&) @ 0x000000000c88dc7d in /usr/lib/debug/usr/bin/clickhouse.debug
7. ./build_docker/./base/poco/Util/src/Application.cpp:0: Poco::Util::Application::run() @ 0x00000000153a4dc6 in /usr/lib/debug/usr/bin/clickhouse.debug
8. ./build_docker/./programs/server/Server.cpp:414: DB::Server::run() @ 0x000000000c889a51 in /usr/lib/debug/usr/bin/clickhouse.debug
9. ./build_docker/./base/poco/Util/src/ServerApplication.cpp:132: Poco::Util::ServerApplication::run(int, char**) @ 0x00000000153adc39 in /usr/lib/debug/usr/bin/clickhouse.debug
10. ./build_docker/./programs/server/Server.cpp:0: mainEntryClickHouseServer(int, char**) @ 0x000000000c88694a in /usr/lib/debug/usr/bin/clickhouse.debug
11. ./build_docker/./programs/main.cpp:0: main @ 0x0000000007210b18 in /usr/lib/debug/usr/bin/clickhouse.debug
12. ? @ 0x00007f4258f54d90
13. ? @ 0x00007f4258f54e40
14. _start @ 0x0000000005f6d6ae in /usr/lib/debug/usr/bin/clickhouse.debug
 (version 24.1.5.6 (official build))
```

**Pull request merged**